pkg/driver/options.go

```diff
@@ -0,0 +1,108 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package driver
+
+import (
+	"fmt"
+	"time"
+
+	flag "github.com/spf13/pflag"
+	cliflag "k8s.io/component-base/cli/flag"
+)
+
+// Options contains options and configuration settings for the driver.
+type Options struct {
+	Mode Mode
+
+	// #### Server options ####
+
+	//Endpoint is the endpoint for the CSI driver server
+	Endpoint string
+	// HttpEndpoint is the TCP network address where the HTTP server for metrics will listen
+	HttpEndpoint string
+	// EnableOtelTracing is a flag to enable opentelemetry tracing for the driver
+	EnableOtelTracing bool
+
+	// #### Controller options ####
+
+	// ExtraTags is a map of tags that will be attached to each dynamically provisioned
+	// resource.
+	ExtraTags map[string]string
+	// ExtraVolumeTags is a map of tags that will be attached to each dynamically provisioned
+	// volume.
+	// DEPRECATED: Use ExtraTags instead.
+	ExtraVolumeTags map[string]string
+	// ID of the kubernetes cluster.
+	KubernetesClusterID string
+	// flag to enable sdk debug log
+	AwsSdkDebugLog bool
+	// flag to warn on invalid tag, instead of returning an error
+	WarnOnInvalidTag bool
+	// flag to set user agent
+	UserAgentExtra string
+	// flag to enable batching of API calls
+	Batching bool
+	// flag to set the timeout for volume modification requests to be coalesced into a single
+	// volume modification call to AWS.
+	ModifyVolumeRequestHandlerTimeout time.Duration
+
+	// #### Node options #####
+
+	// VolumeAttachLimit specifies the value that shall be reported as "maximum number of attachable volumes"
+	// in CSINode objects. It is similar to https://kubernetes.io/docs/concepts/storage/storage-limits/#custom-limits
+	// which allowed administrators to specify custom volume limits by configuring the kube-scheduler. Also, each AWS
+	// machine type has different volume limits. By default, the EBS CSI driver parses the machine type name and then
+	// decides the volume limit. However, this is only a rough approximation and not good enough in most cases.
+	// Specifying the volume attach limit via command line is the alternative until a more sophisticated solution presents
+	// itself (dynamically discovering the maximum number of attachable volume per EC2 machine type, see also
+	// https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/347).
+	VolumeAttachLimit int64
+	// ReservedVolumeAttachments specifies number of volume attachments reserved for system use.
+	// Typically 1 for the root disk, but may be larger when more system disks are attached to nodes.
+	// This option is not used when --volume-attach-limit is specified.
+	// When -1, the amount of reserved attachments is loaded from instance metadata that captured state at node boot
+	// and may include not only system disks but also CSI volumes (and therefore it may be wrong).
+	ReservedVolumeAttachments int
+}
+
+func (o *Options) AddFlags(f *flag.FlagSet) {
```

Comment:
To avoid changing the current behavior, this should check against `o.Mode` and only apply the options related to the currently selected mode

Reply:
Discussed internally to clarify request (I initially misunderstood as adding additional validation to main entry).

Fixed:

```
./aws-ebs-csi-driver controller --volume-attach-limit=123     
unknown flag: --volume-attach-limit
```